### DISCUSSION TITLE
freebsd

### DISCUSSION THREAD

**Discussion by joske** [aristocratos/btop#113] [General]
@aristocratos You probably noticed I started work on FreeBSD port ;-). Sorry to steal your work. It went surprisingly well, it's already fully functional. Everything can be done with syscalls. 

Still to do:
* mem leak in processes (like osx) - rest seems fine
* cpu % wonky
* CPU temperature
* Makefile is hacked at the moment, still need to separate with ifeq etc

**Comment by aristocratos** [comment]
@joske 
Actually hadn't noticed, nice work! :)

> Sorry to steal your work.

No worries man, just means I get more time to work on the core functionality of the whole program.

> mem leak in processes (like osx) - rest seems fine

The cpu percent for processes is also not working correctly in OSX, need to look it over.

> Makefile is hacked at the moment, still need to separate with ifeq etc

Will take a look at the freebsd code tomorrow and fix up the Makefile and see if I can find a solution to the leak(s) in Proc.

Are you using any custom gcc toolchain or just standard tools and the gcc11 package?

**Comment by joske** [comment]
Do you have a FreeBSD system already? If not, I recommend using ghostbsd.
Has graphical installer and boots directly in desktop like a Linux distro.
With FreeBSD proper you spend an evening trying to get the desktop to work
🙄. I installed FreeBSD on virtualbox. Feels like Linux in 1996 when I was
starting out... Then I remembered an old laptop and installed ghostbsd on
that. Much impressed with that. Even had working wifi and sleep support ☺️.

I just installed gcc11, yes. To get vscode to work, there's also some
manual steps to get it to code complete etc.

**Comment by aristocratos** [comment]
I have an old VM setup with ghostbsd for testing bpytop, should be fine after some updates.

> With FreeBSD proper you spend an evening trying to get the desktop to work
🙄. I installed FreeBSD on virtualbox. Feels like Linux in 1996 when I was
starting out..

Yeah, I remember installing slackware from some insane number of floppies in the distant past, not an experience I wanna relive :P

**Comment by aristocratos** [comment]
@joske 
Fixed the leaking, added a RAII wrapper for `kvm_openfiles` since it needed a `kvm_close` to free up allocations.
And fixed tree mode by removing pid 0 for kernel process and pseudo process `idle`.

The cpu percentages for processes is indeed a bit wonky. Will take another look at it when I've got freebsd setup on some real hardware.
I noticed vscodes remote ssh targets didn't support freebsd, so is currently working on the code in a remote folder without any intellisense or linting, not ideal.
But have laptop I'm gonna set up a fresh install of ghostbsd on and run vscode directly on the machine instead.

Awesome work otherwise, alot of stuff seems to be working nicely :)

**Comment by joske** [comment]
Install the clangd extension, then you have at least basic auto complete.
Need to set the path to clangd to clangd12. I developed directly on the
FreeBSD laptop. But the intellisense plugin doesn't work like it does on
Linux 😕

Great! I was working on CPU temperature, it seemed simple but somehow what
I get back from the sysctl is 3100 degrees 🤔. Sysctl on the command line
returns same values.

**Comment by aristocratos** [comment]
> Install the clangd extension, then you have at least basic auto complete.

Won't work from remote since it will look for the system header files on the local system and not the remote freebsd system.
Don't know why they haven't added freebsd as a valid remote ssh target yet, since electron and vscode in general seems to work fine on freebsd...

But shouldn't be an issue when I've got it setup on some real hardware.

> Great! I was working on CPU temperature, it seemed simple but somehow what
I get back from the sysctl is 3100 degrees 🤔. Sysctl on the command line
returns same values.

Multiplied by 100 for int to float conversion possibly? Would be weird though since it's usually a multiplier by 1000 or more in those cases.

Is cpu hz working btw? Don't get any value in the VM.

**Comment by joske** [comment]
No it's not multiplied by 100. Like completely different from actual value.
Will investigate when I have time, no time anymore tonight.

Indeed it doesn't show although it was implemented. Will check that too.

Battery is working.

**Comment by aristocratos** [comment]
Could maybe find something useful looking at [hwstat](https://www.freshports.org/sysutils/hwstat), can't seem to find the actual source files for it though...

**Comment by joske** [comment]
found the solution on a freebsd mailinglist from 2007...

[image: image.png]

**Reply by aristocratos** [reply]
What was the solution? The image you posted is not loading.

**Reply by aristocratos** [reply]
Nvm, saw the comment in the commit.

> and offset to absolute zero

Now that's just weird :)

**Comment by joske** [comment]
only thing missing is now disk io...

**Comment by aristocratos** [comment]
Fixed up the Makefile and made it even with the OSX Makefile, should make it easier when it's all finally merged.

Builds without any issues now with the OPTFLAGS from main as long as `coreutils` and `gmake` is installed. Don't know what's up with freebsd's builtin `make`, errors on every odd line :\

**Comment by joske** [comment]
Tried to find a way to find disk io stats per mountpoint (like linux/osx), but it doesn't seem possible. I can get stats per device (looking at iostat.c from FreeBSD), but as my system is using zfs, you don't even know from getmntinfo which device the mount comes from :-/. 

You get only 1 entry back form devstat: ada (the only disk in the machine), plus I fetch/print stats for bytes read/written.

I will commit my devstat_* code, hopefully you find a way to make it useful (maybe only show per disk stats in io panel? don't know how to do that).

Edit: here is a link to iostat.c : 

https://gitlab.unc.edu.ar/nicolaspapp/freebsd/-/blob/stable/12/usr.sbin/iostat/iostat.c

**Comment by joske** [comment]
@aristocratos I've added a few commits to freebsd branch, no more mem leak, and disk IO is also fixed for all mountpoints (at least for ZFS). 
I think this is now ready to be merged. As this branch is very old (and has many conflicts with main), I'd propose to just take the btop_collect.cpp file from the freebsd folder. So I didn't create a PR for this.

**Reply by joske** [reply]
@aristocratos scratch that last bit, there are some other changes, I'll attempt to merge main into freebsd.

**Comment by aristocratos** [comment]
@joske 
> I've added a few commits to freebsd branch, no more mem leak, and disk IO is also fixed for all mountpoints

Oh, nice!

Have been taking a little break, will look over the PR later this week, some changes in the process sorting that needs updating in the collect for freebsd and I believe the process cpu usage is still incorrect, but I believe I know how to fix that.
But all that can be added after the merge including documentation in README.md for FreeBSD.

**Reply by aristocratos** [reply]
PR approved, didn't think the merge would be that clean :P

**Reply by joske** [reply]
Merged. I worked until after midnight on getting it in this state ;-)

**Comment by tessus** [comment]
I am running OPNsense which is FreeBSD 13.1. However, I get the following message, when I try to install btop:

```
pkg: No packages available to install matching 'btop' have been found in the repositories
```

I am very new to FreeBSD, so I am sorry for asking such a weird question. How do I install btop on FreeBSD (in this case OPNsense). I don't want to compile it on my FreeBSD router/firewall, and I don't even know whether the necessary dev packages can be installed either.

**Reply by joske** [reply]
Not a BTOP problem. Ask OPNsense to package it.

**Reply by tessus** [reply]
Interesting, will do. I thought it would still use the FreeBSD repos for not OPNsense specific packages. I guess I was wrong. Thanks.
